src/Middleware/RequestThrottling/src/Internal/RequestQueue.cs

```diff
@@ -9,36 +9,40 @@ namespace Microsoft.AspNetCore.RequestThrottling.Internal
 {
     internal class RequestQueue : IDisposable
     {
-        private SemaphoreSlim _semaphore;
+        private readonly int _maxConcurrentRequests;
+        private readonly int _requestQueueLimit;
+        private readonly SemaphoreSlim _semaphore;
+
         private object _waitingRequestsLock = new object();
-        public readonly int MaxConcurrentRequests;
-        public int WaitingRequests { get; private set; }
+        private int _waitingRequests;
 
-        public RequestQueue(int maxConcurrentRequests)
+        public RequestQueue(int maxConcurrentRequests, int requestQueueLimit)
         {
-            MaxConcurrentRequests = maxConcurrentRequests;
+            _maxConcurrentRequests = maxConcurrentRequests;
+            _requestQueueLimit = requestQueueLimit;
             _semaphore = new SemaphoreSlim(maxConcurrentRequests);
         }
 
-        public async Task EnterQueue()
+        public async Task<bool> TryEnterQueueAsync()
         {
-            var waitInQueueTask = _semaphore.WaitAsync();
-
-            var needsToWaitOnQueue = !waitInQueueTask.IsCompletedSuccessfully;
-            if (needsToWaitOnQueue)
+            lock (_waitingRequestsLock)
             {
-                lock (_waitingRequestsLock)
+                if (_waitingRequests >= _requestQueueLimit)
```

Comment:
So a _requestQueueLimit of 0 still breaks this, right? _waitingRequests will be 0 and all requests will be rejected even if the _semaphore has space.

That's fine, but we'd want to put argument validation on _requestQueueLimit.

Reply:
@DylanDmitri and I discussed this yesterday. The fix will be to check both _waitingRequests and concurrentRequests to both limits. Specifically: `_waitingRequests + ConcurrentRequests >= _requestQueueLimit + _maxConcurrentRequests`

Reply:
That would help. Except ConcurrentRequests isn't controlled under the lock, it's controlled by the _semaphore so your results are going to be inconsistent. e.g. I can get 10k requests past this lock if they pause before acquiring the semaphore. It's unlikely to get that off balance, but it will certainly drift.

I expect this will need to abandon the semaphore soon, it's already outgrowing it.

Reply:
> it's controlled by the _semaphore so your results are going to be inconsistent.

Bleh you are right. I didn't realize we used _semaphore.CurrentCount rather than counting outside of the semaphore.

Reply:
We could manage our own count independently of the semaphore for now. Our counter would increment before and decrement after the semaphore's internal counter does, that way you'd still never let through more than `_requestQueueLimit + _maxConcurrentRequests` through at a time.

Reply:
I'll try three different implementations and then see what feels best:
- no semaphores just locks and counts
- two semaphores, one for queue one for server
- something involving a concurrent queue

Reply:
wow the first one is bad defining a queue yourself is not the way to go

Reply:
I am feeling tied to the log messages I already defined; in order to log the proper stuff in the proper places there's considerable gymnastics. Will revise which types of messages exist in these 'experiment' branches.

Reply:
> I am feeling tied to the log messages I already defined; in order to log the proper stuff in the proper places there's considerable gymnastics.

Consider that it's not just logging. Going forward, we'll probably want to increment/decrement performance counters in the same places, potentially call user-defined callbacks, etc...

Reply:
Making ConcurrentRequests a local field rather than checking _semaphore.CurrentCount makes sense. Just make sure you add another lock in Release when derecementing it.